### PR TITLE
ccxt backtest and ccxt data cache, ccxt backtest example upload, _strategy.py, ccxt.py modified for ccxt backtest

### PR DESCRIPTION
**Pull request by mrchaos**
● ccxt data cache :  lumibot/tools/ccxt_data_store.py
 - This module used duckdb to cache the ccxt data, see the in-source documentation for details.
 - Used when backtesting CCXT.
 
● ccxt data cache test module: tests/test_ccxt_store.py
 
● ccxt backtesting data source : lumibot/data_sources/ccxt_backtesting_data.py
 - This module is similar to the structure of yahoo_data.py.
 - The ccxt data cache module is used here.
 
● ccxt backtesting : lumibot/backtesting/ccxt_backtesting.py
 - This module is similar to the structure of yahoo_backtesting.py.

● ccxt backtesting example : lumibot/example_strategies/ccxt_backtesting_example.py
 - binance, bitmex backtesting example
 - Use `ccxt backtesting data source` as the benchmark data source.

● _strategy.py, ccxt.py  modified for ccxt backtesting

### CODE REVIEW

**Review by grzesir — CHANGES_REQUESTED** [review]
This looks fantastic! Great work!

One small thing: can you add a test for the example strategy? Every example strategy has a test in this file: https://github.com/Lumiwealth/lumibot/blob/dev/tests/backtest/test_example_strategies.py

It should be very straight forward, but it will make sure that future code does not break this backtesting capability (and sometimes helps reveal issues)

File: tests/test_ccxt_store.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+from lumibot.tools import CcxtCacheDB
+import pytest
+import duckdb
```
Comment (grzesir) [inline tests/test_ccxt_store.py:3]: this should be a dependency in setup.py

Comment (mrchaos) [reply tests/test_ccxt_store.py:3]: I added the dependency to setup.py.

**Comment by mrchaos** [comment]
Added the example strategy unit test code to the test_example_strategies.py file.

**Comment by mrchaos** [comment]
> @mrchaos
> 
> This looks fantastic! Great work!
> 
> One small thing: can you add a test for the example strategy? Every example strategy has a test in this file: https://github.com/Lumiwealth/lumibot/blob/dev/tests/backtest/test_example_strategies.py
> 
> It should be very straight forward, but it will make sure that future code does not break this backtesting capability (and sometimes helps reveal issues)

Added the example strategy unit test code to the test_example_strategies.py file.

**Comment by grzesir** [comment]
> Added the example strategy unit test code to the test_example_strategies.py file.

thank you @mrchaos ! this is a great addition to lumibot, you will now be remembered in history as the one who added this feature.

btw we should probably add this to the docs too, I can do this if you write out a quick overview of how it works.

for now im going to merge this and it will be deployed in our next release (probably over the weekend)

**Comment by grzesir** [comment]
@mrchaos merged.

please let me know what we should put in the docs for this. if you give me a short writeup (a few sentences with example code) i can add it in

**Comment by mrchaos** [comment]
> @mrchaos merged.
> 
> please let me know what we should put in the docs for this. if you give me a short writeup (a few sentences with example code) i can add it in

I'll send you the exmaple code and documentation after some more testing.

**Comment by grzesir** [comment]
Ok perfect, thank you.

Robert Grzesik 347-635-3416

**Pull request merged**